rework the revocation guide (#46)
biscuit-auth/website: content/docs/guides/revocation.md

[extra]
lead = "How to revoke a token"
toc = true
top = false
+++

## Revocation identifiers

Once a token has been created, we need a way to invalidate it, either due to its lifecycle,
like logging out or decommissioning a service, or because it was compromised. Tokens can come with
an expiration date, but they are not sufficient, as there will always be some delay between the leak
and the expiration date. So we need a way to revoke currently existing tokens.

> Revoking bearer tokens lik biscuits is usually done through revocation lists: a list of

Reviewer: ```suggestion
Revoking bearer tokens like biscuits is usually done through revocation lists: a list of
```

tokens that are no longer accepted is shared with all verifying parties. When authorizing a biscuit token,
the library will look it up in the list and refuse the request if it finds it.

Such a mechanism relies on being able to uniquely identify tokens: we want to revoke only the tokens
that are not valid anymore, without affecting other tokens (even tokens with the same content that have been issued
to another holder). With offline attenuation, biscuits introduce another constraint: revoking a token should also
revoke all derived tokens (else it would be trivial to circument revocation).

The biscuit spec (and libraries) provides you with:

 - a way to uniquely identify tokens (two biscuits with the same payload and secret key will be different)
 - a way to identify groups of tokens derived from the same parent token
 - a way to reject tokens based on that identification during authorization

Biscuit's revocation identifiers are unique and generated directly from the token's structure, there is no need to add

> them explicitely, as would be done with Macaroons or the  "jti" claim in JWT.

Reviewer: ```suggestion
them explicitely, as would be done with Macaroons or the  `jti` claim in JWT.
```

😎

know a list of tokens that must be refused, and that list changes dynamically, so we are
reintroducing some state in the system.

But revocation has properties that make it nice to implement in stateless architectures:
- we do not need to know about all of the tokens, only those that were revoked, which will
be much smaller
- the list of revoked tokens will naturally grow, but if tokens have an expiration date, they
can be purged from the list after a while
- it is read-oriented and highly cacheable: once a token was added to the revocation list,
we won't modify its entry (except when purging), so we don't need synchronization or consensus
- revocation lists do not hold any critical or private information, they can be shared with
every service

So handling revocation is adding some shared state, but much more limited than what we would
have with a fully centralized architecture.

## How to implement revocation in our infrastructure?

We need a reliable way to transmit revocation information to services. That will depend on
how quickly we want to disseminate it, and how much complexity we can bear.

> ### The basic solution: download the revocation list at startup

Reviewer: ```suggestion
### The basic solution: read the revocation list at startup
```

> In some cases, like communication betwen automated services, revocation is rare, mostly when

Reviewer: ```suggestion
In some cases, like communication between automated services, revocation is rare, mostly when
```

> a service is stopped, so the revocation list is mostly static and small. If we can accept some

Reviewer: ```suggestion
a service is stopped, or a token is leaked, so the revocation list is mostly static and small. If we can accept some
```

> delay in synchronization, we can have services download the revocation list at startup. They

Reviewer: ```suggestion
manual operations, and a (slight) delay in synchronization, we can have services read the revocation list at startup. They
```

will check tokens from an in memory list, that will stay the same for the entire life of the

> service.

Reviewer: ```suggestion
service (until it is restarted).
```

The tradeoff here is that if we need to revoke a token urgently, we will need to redeploy
a lot of services at once.

> 

Reviewer: ```suggestion

In the case where there is only one service accepting tokens, the revocation list can be read from config (a config file or environment variables).

```

> Since the revocation list is small and static, it can be stored as a file in an object

Reviewer: ```suggestion
In the case where more services accept tokens, it will become necessary to have a centrally defined list that is then distributed to all services.
Since the revocation list is small and static, it can be stored as a file in an object
```

store like S3, and downloaded via HTTP. That file can be updated independently whenever

> a service stops, or when one of the token expires.

Reviewer: ```suggestion
a service stops, or when one of the token expires or is leaked.

Depending on how quick you want revocation to take effect, you can either wait for the services to restart or trigger a restart of all affected services.
```

### Slightly more advanced: download the revocation list regularly

The natural next step from the previous solution: instead of downloading the list once, it

> is downloaded regularly tokeep it up to date. There is still a gap between revocation and

Reviewer: ```suggestion
is downloaded regularly to keep it up to date. There is still a gap between revocation and
```

top = false
+++

## Why should we plan for token revocation?

There are two competing approaches to session management in authorization, that will drive

> architectural decisions:

Reviewer: ```suggestion
architectural decisions:

```

Author: apparently zola (and now github too) can handle lists without an empty line before, and this has no effect on the rendered text

Reviewer: Yeah it's an open discussion https://talk.commonmark.org/t/blank-lines-before-lists-revisited/1990/35

- in *stateful* systems, all authorizations are performed through one service or database

> that olds the list of currently active sessions

Reviewer: ```suggestion
that holds the list of currently active sessions
```

- in *stateless* systems, authorization can be performed independently in any service, only
using information from the token and the service. In particular, the service cannot know
about all of the currently active sessions

Those two solutions are often compared on their ability to close a session. Why? Can't we
just set an expiration date? Even with expiration date we would still need a way to close

> a session, to implement the log out functionality. That feature is comon, expected by users,

Reviewer: ```suggestion
a session, to implement the log out functionality. That feature is common, expected by users,
```

and needed in multiple situations (public computer, disconnecting sessions from a stolen
phone...)?

In stateful systems, closing a session is easy: delete the session's information from the
database and that's it. In stateless systems, this is more complex: how do we make sure
all services know that the session is invalid? That means reintroducing some shared state,
so is the stateless design impossible after all? Shouldn't we go back to stateful systems?

If the architecture we are designing can rely on a central state, it will be the simplest
approach and probably the right solution. But there are good reasons to choose the stateless

> design:

Reviewer: ```suggestion
design:

```

- **scaling**: a central authorization service that is queried on every request is a single point
of failure for the entire system. If it is down, nobody can log in, but existing sessions will
fail too. A stateless approach will decouple session creation from authorization, so existing
sessions can still work when the authentication service fails.
- **isolation**: the service receiving the request might be less trusted and should not be
able to access session information.
- **authentication delegation**: authentication could be in a separate service (example: SSO)
that can't be queried on every request. That service could even be managed by a different
company.

In those cases, separating authorization from session creation makes sense, but then how do
we close a session? It is usually done through token revocation: the authorizer needs to
know a list of tokens that must be refused, and that list changes dynamically, so we are
reintroducing some state in the system.

> But revocation has properties that make it nice to implement in stateless architectures:

Reviewer: ```suggestion
But revocation has properties that make it nice to implement in stateless architectures:

```

- we do not need to know about all of the tokens, only those that were revoked, which will
be much smaller
- the list of revoked tokens will naturally grow, but if tokens have an expiration date, they
can be purged from the list after a while

> - it is read oriented and highly cacheable: once a token was added to the revocation list,

Reviewer: ```suggestion
- it is read-oriented and highly cacheable: once a token was added to the revocation list,
```

If the architecture we are designing can rely on a central state, it will be the simplest
approach and probably the right solution. But there are good reasons to choose the stateless
design:
- **scaling**: a central authorization service that is queried on every request is a single point
of failure for the entire system. If it is down, nobody can log in, but existing sessions will
fail too. A stateless approach will decouple session creation from authorization, so existing
sessions can still work when the authentication service fails.
- **isolation**: the service receiving the request might be less trusted and should not be
able to access session information.
- **authentication delegation**: authentication could be in a separate service (example: SSO)
that can't be queried on every request. That service could even be managed by a different
company.

In those cases, separating authorization from session creation makes sense, but then how do
we close a session? It is usually done through token revocation: the authorizer needs to
know a list of tokens that must be refused, and that list changes dynamically, so we are

> reintroducing some state in the system.

Reviewer 2: This sentence is a repetition. From my Monkey POV, It is implicit hear

Author: the tricky parts need a lot of repetition :)

Reviewer: 👍

We need a reliable way to transmit revocation information to services. That will depend on
how quickly we want to disseminate it, and how much complexity we can bear.

### The basic solution: read the revocation list at startup

In some cases, like communication between automated services, revocation is rare, mostly when
a service is stopped or a token is leaked, so the revocation list is mostly static and small. If we can accept some
manual operations, and a (slight) delay in synchronization, we can have services read the revocation list at startup. They
will check tokens from an in memory list, that will stay the same for the entire life of the
service (until it is restarted).

The tradeoff here is that if we need to revoke a token urgently, we will need to redeploy
a lot of services at once.

In the case where there is only one service accepting tokens, the revocation list can be read from config (a config file or environment variables).

In the case where more services accept tokens, it will become necessary to have a centrally
defined list that is then distributed to all services.
Since the revocation list is small and changes rarely, it can be stored as a file in an object

> store like S3, and downloaded via HTTP. That file can be updated independently whenever

Reviewer 2: This is what Terraform do when dealing with sharing state. But it adds a KV store for locking resource for concurrent access

Author: the nice thing here is that it does not need much locking, there won't be concurrent writes

indefinitely.

### OAuth specific usage

In OAuth based systems, API clients hold an access token, used to query the API, and a refresh
token, used to get a new access token. The idea here is that the access token is used often
and potentially on less trusted services, so it has a short expiration, while the refresh
token has a longer lifetime because it is only used once in a while, and only with the
authorization server.

While it is common to see applications with a permanent refresh token, this will causes
issues with the revocation list, causing it to grow, and current practices evolved
towards a different approach.

It is now recommended to have a refresh token with an expiration date, that can be long,
and have that refresh token be single use. When it is sent to the authorization server
to get a new access token, the authorization server will revoke the old refresh token
and issue new refresh and access tokens. The interesting property here is that if the
authorization server sees the same refresh token twice, it means that the token was
stolen: either the thief or the legitimate client already used the refresh token, and

> the other one is ow requesting an access token too. In that case, the authorization

Reviewer: ```suggestion
the other one is now requesting an access token too. In that case, the authorization
```
